### PR TITLE
Bluetooth: Move crypto toolbox functions from `smp.c` into their own file

### LINKED ISSUE #51297: Bluetooth: Implement H8 function from cryptographic toolbox
@kruithofa and @cvinayak are requesting that someone implements the Cryptographic Toolbox function "H8" (as defined in BT Core ref Vol 6, part E, 1.1.1) and associated test for it. It is needed by the Controller to derive the GSK used in encrypting the broadcast ISA data.

https://github.com/zephyrproject-rtos/zephyr/issues/51297#tasklist-block-d5630141-6313-4144-bef9-54cccea3933c

### PR DESCRIPTION
**Pull request by theob-pro**
Fix #51297

### CODE REVIEW

**Review by carlescufi — CHANGES_REQUESTED** [review]
Any reason not to use the more obvious (and short) `bt_crypto.c`?

**Comment by alwa-nordic** [comment]
> Any reason not to use the more obvious (and short) `bt_crypto.c`?

There is a file `bluetooth/controller/crypto/crypto.c`. That could be confusing. Also `crypto_toolbox` in a reference to the Bluetooth specification that defines these functions in sections named "Crypto Toolbox".

File: drivers/bluetooth/CMakeLists.txt [hunk 1]
```diff
@@ -3,3 +3,7 @@
 zephyr_library()
 zephyr_library_include_directories(${ZEPHYR_BASE}/subsys/bluetooth)
 add_subdirectory(hci)
+
+if(${CONFIG_BT_NO_DRIVER})
```
Comment (jori-nordic) [inline drivers/bluetooth/CMakeLists.txt:7]: why is this needed?

File: subsys/bluetooth/crypto_toolbox/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_include_directories(
+    include
+)
+
+zephyr_library()
+
+target_include_directories(${ZEPHYR_CURRENT_LIBRARY} PUBLIC
```
Comment (jori-nordic) [inline subsys/bluetooth/crypto_toolbox/CMakeLists.txt:9]: I think L3 and this are redundant. Can you remove one of them?

File: subsys/bluetooth/crypto_toolbox/crypto_toolbox.c [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+}
+
+int bt_crypto_toolbox_f4(const uint8_t *u, const uint8_t *v, const uint8_t *x, uint8_t z,
+			 uint8_t res[16])
+{
+	uint8_t xs[16];
+	uint8_t m[65];
+	int err;
+
+	/*
```
Comment (jori-nordic) [inline subsys/bluetooth/crypto_toolbox/crypto_toolbox.c:62]: can you add back the `BT_DBG` statements that were in `smp.c`'s version?

File: subsys/bluetooth/crypto_toolbox/include/crypto_toolbox.h [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+/**
+ * @brief Cryptographic Toolbox g2
+
+ * Defined in Core Vol. 3, part H 2.2.9.
+ *
+ * @param u 256-bit
+ * @param v 256-bit
+ * @param x 128-bit
+ * @param y 128-bit
+ * @param[out] passkey
```
Comment (jori-nordic) [inline subsys/bluetooth/crypto_toolbox/include/crypto_toolbox.h:84]: please also use `param[in]` for consistency.

File: subsys/bluetooth/crypto_toolbox/include/crypto_toolbox.h [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+/**
+ * @brief Cryptographic Toolbox h6
+ *
+ * Link key conversion defined in Core Vol. 3, part H 2.2.10.
+ *
+ * @param w 128-bit key
+ * @param key_id 32-bit
+ * @param[out] res 128-bit
+ * @retval 0 Computation was successful. @p res contains the result.
+ * @retval -EIO
```
Comment (jori-nordic) [inline subsys/bluetooth/crypto_toolbox/include/crypto_toolbox.h:100]: can you add a reason for that error code? e.g. `@retval -EIO Computation failed.`

File: subsys/bluetooth/host/smp.c [hunk 6]
```diff
@@ -512,34 +509,6 @@ static struct net_buf *smp_create_pdu(struct bt_smp *smp, uint8_t op, size_t len
...
-	}
-
-	if (tc_cmac_final(out, &state) == TC_CRYPTO_FAIL) {
-		return -EIO;
-	}
-
-	return 0;
-}
-
 static int smp_d1(const uint8_t *key, uint16_t d, uint16_t r, uint8_t res[16])
```
Comment (jori-nordic) [inline subsys/bluetooth/host/smp.c:512]: I guess you can now move `d1` next to its sole caller.

File: tests/bluetooth/crypto_toolbox/Kconfig [hunk 7]
```diff
@@ -0,0 +1,9 @@
+config BT_CRYPTO_TOOLBOX_TEST
+    bool
+    default y
+    select BT_CRYPTO_TOOLBOX
+    help
```
Comment (jori-nordic) [inline tests/bluetooth/crypto_toolbox/Kconfig:5]: don't need a helptext since it's just a shim/hack to select the non-user-facing `BT_CRYPTO_TOOLBOX`.
I would name it `BT_ENABLE_CRYPTO_TOOLBOX` so it's clear what it does.

**Review by jori-nordic — CHANGES_REQUESTED** [review]
a few small comments, but otherwise looks good.

**Comment by Thalley** [comment]
> > Any reason not to use the more obvious (and short) `bt_crypto.c`?
> 
> There is a file `bluetooth/controller/crypto/crypto.c`. That could be confusing. Also `crypto_toolbox` in a reference to the Bluetooth specification that defines these functions in sections named "Cryptographic Toolbox".

We actually have 3 files called crypto.c (in just the Bluetooth subsystem):
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/crypto.c
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/crypto.c
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/controller/crypto/crypto.c

While I don't disagree with you, it's not a very solid argument since it already happened 3 times :D

**Review by alwa-nordic — DISMISSED** [review]

File: drivers/bluetooth/CMakeLists.txt (hunk 1 above)
Comment (carlescufi) [inline drivers/bluetooth/CMakeLists.txt:7]: @tejlmand FYI

Comment (tejlmand) [reply drivers/bluetooth/CMakeLists.txt:7]: it is not wrong, but the `${}` can be omitted.

And as this is generally what we do, I recommend changing to:
```suggestion
if(CONFIG_BT_NO_DRIVER)
```

File: subsys/bluetooth/crypto_toolbox/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library()
+
+zephyr_include_directories(
```
Comment (tejlmand) [inline subsys/bluetooth/crypto_toolbox/CMakeLists.txt:5]: minor nit: there is no need for a line wrap here and below.

File: tests/bluetooth/crypto_toolbox/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+cmake_minimum_required(VERSION 3.20.0)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(foo)
+
+target_sources(app PRIVATE
+  src/test_bt_crypto_toolbox.c
+)
+
+target_link_libraries(app PRIVATE
+  subsys__bluetooth__crypto_toolbox
```
Comment (tejlmand) [inline tests/bluetooth/crypto_toolbox/CMakeLists.txt:12]: why is this needed ?

The app lib is inside the whole archive flag, as are Zephyr libraries such as `subsys__bluetooth__crypto_toolbox`.

File: tests/bluetooth/crypto_toolbox/Kconfig [hunk 10]
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2022 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+config BT_ENABLE_CRYPTO_TOOLBOX
+	bool
+	default y
+	select BT_CRYPTO_TOOLBOX
```
Comment (tejlmand) [inline tests/bluetooth/crypto_toolbox/Kconfig:7]: we are in a dedicated test sample, so why have this as a `select BT_CRYPTO_TOOLBOX` on not directly as a `CONFIG_BT_CRYPTO_TOOLBOX=y` in the prj.conf ?

**Review by tejlmand — CHANGES_REQUESTED** [review]
some minor nits and questions.

File: drivers/bluetooth/CMakeLists.txt (hunk 1 above)
Comment (carlescufi) [reply drivers/bluetooth/CMakeLists.txt:7]: Thanks @tejlmand. @theob-pro please do make the change.

File: tests/bluetooth/crypto_toolbox/CMakeLists.txt (hunk 9 above)
Comment (alwa-nordic) [reply tests/bluetooth/crypto_toolbox/CMakeLists.txt:12]: I think it fails to compile without this. What is the correct way to get this lib in?

Comment (theob-pro) [reply tests/bluetooth/crypto_toolbox/CMakeLists.txt]: I was thinking the same but I tried to remove it and there is no problem.

**Comment by carlescufi** [comment]
@theob-pro still blocking this since you haven't changed the name.

**Comment by carlescufi** [comment]
> > > Any reason not to use the more obvious (and short) `bt_crypto.c`?
> > 
> > 
> > There is a file `bluetooth/controller/crypto/crypto.c`. That could be confusing. Also `crypto_toolbox` in a reference to the Bluetooth specification that defines these functions in sections named "Cryptographic Toolbox".
> 
> We actually have 3 files called crypto.c (in just the Bluetooth subsystem): https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/crypto.c https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/crypto.c https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/controller/crypto/crypto.c
> 
> While I don't disagree with you, it's not a very solid argument since it already happened 3 times :D

Exactly. What exactly is the problem with files being named identically if they don't share a common path?

File: tests/bluetooth/bt_crypto/Kconfig [hunk 11]
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2022 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+config BT_ENABLE_CRYPTO
```
Comment (carlescufi) [inline tests/bluetooth/bt_crypto/Kconfig:4]: Why is this Kconfig option needed?

**Review by carlescufi — CHANGES_REQUESTED** [review]

File: tests/bluetooth/bt_crypto/Kconfig (hunk 11 above)
Comment (theob-pro) [reply tests/bluetooth/bt_crypto/Kconfig:4]: Because `BT_CRYPTO` cannot be selected by the user, it can't be added in `prj.conf`.

File: drivers/bluetooth/CMakeLists.txt [hunk 12]
```diff
@@ -3,3 +3,7 @@
 zephyr_library()
 zephyr_library_include_directories(${ZEPHYR_BASE}/subsys/bluetooth)
 add_subdirectory(hci)
+
+if(CONFIG_BT_NO_DRIVER)
+    zephyr_library_property(ALLOW_EMPTY TRUE)
```
Comment (tejlmand) [inline drivers/bluetooth/CMakeLists.txt:8]: why is this needed ?

It should only be used on old libraries where the amount of work needed to clean up the CMake code was considered to excessive compared to the gain.

Please see this commit message:
https://github.com/zephyrproject-rtos/zephyr/commit/153196b3784032b9dc706c9e2ac3a25e1aedfdbb

> In general libraries should not be created unless there are files to
places inside of it, however in some cases there are so many
`zephyr_library_sources_ifdef(<setting> <file.c>)`
where testing each setting before creating the library may not be
desired.

File: tests/bluetooth/bt_crypto/Kconfig (hunk 11 above)
Comment (tejlmand) [reply tests/bluetooth/bt_crypto/Kconfig:4]: but then you can just do:
```
config BT_CRYPTO
	default y
```

no need for a new Kconfig.

File: tests/bluetooth/bt_crypto/Kconfig [hunk 13]
```diff
@@ -0,0 +1,10 @@
...
+	bool
+	default y
+	select BT_CRYPTO
```
Comment (tejlmand) [inline tests/bluetooth/bt_crypto/Kconfig:7]:
```suggestion
config BT_CRYPTO
	default y
```

**Review by tejlmand — CHANGES_REQUESTED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by jori-nordic — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]
lgtm

**Comment by sjanc** [comment]
#AutoPTS run zephyr SM

**Comment by codecoup-tester** [comment]
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/51809#issuecomment-1308711393 after 13:59:57.

**Comment by codecoup-tester** [comment]
AutoPTS Bot results:
Failed tests:
SM      SM/CEN/KDU/BI-02-C              INCONC - ERRATA TSE19199
SM      SM/CEN/KDU/BI-03-C              INCONC - ERRATA TSE19199

**Review by sjanc — APPROVED** [review]
